BUG 2276565: match csi-addons subscription channel to ocs-client-operator

rohan47 opened the pull request:
this change will match csi-addons subscription channel with ocs-client-operator in 4.15

Madhu-1 commented on controllers/clusterversion_controller.go:
```diff
@@ -436,3 +445,40 @@ func (c *ClusterVersionReconciler) ensureConsolePlugin() error {
+	if err != nil {
+		return nil, fmt.Errorf("failed to list subscriptions for %s using labels: %v", subscriptionLabelKey, err)
+	}
+	if len(subscriptions.Items) == 1 {
```
instead of label can we go with known name which we use for subscription?

iamniting replied:
The name can be different, But the package name is unique identity under the spec and we use that in the odf.

Madhu-1 commented on controllers/clusterversion_controller.go:
```diff
@@ -436,3 +445,40 @@ func (c *ClusterVersionReconciler) ensureConsolePlugin() error {
 	return nil
 }
+
+func (c *ClusterVersionReconciler) reconcileCSIAddonsSubscription() error {
```
dont we need to watch for subscription as well?

rohan47 replied:
I am just fetching the subscription and changing the channel if it doesn't match ocs-client-operator as my intention here was to change the channel for csi-addons if it doesn't get upgraded

iamniting replied:
We should watch for subscription if not already, otherwise how this code will be triggered automatically if there is a change in the subscription

rohan47 replied:
added watch for subscriptions, verifying the code changes

Madhu-1 commented on controllers/clusterversion_controller.go:
```diff
@@ -54,7 +56,9 @@ const (
 	// openshift cluster.
-	clusterVersionName = "version"
+	clusterVersionName                          = "version"
+	csiAddonsSubscriptionLabelKeyPrefix         = "operators.coreos.com/odf-csi-addons-operator"
+	ocsClientOperatorSubscriptionLabelKeyPrefix = "operators.coreos.com/ocs-client-operator"
```
who is adding the labels? can we go with names as we know it already.

rohan47 replied:
The labels are getting added by olm as per my understanding
and regarding known name, the name for csi-addons is `odf-csi-addons-operator-stable-4.15-redhat-operators-openshift-marketplace` where the channel part gets included so its simple to use the label

rohan47 replied:
making the change to package name instead of labels as Nitin suggested

Madhu-1 reviewed (commented):
@rohan47 i think the final decision was this is the fix going to be present in all version we don't need to send this PR to 4.15 isn't it?

leelavg commented:
> @rohan47 i think the final decision was this is the fix going to be present in all version we don't need to send this PR to 4.15 isn't it?
- yes, we do need a change in 4.16, even though the final outcome will be same the codepath diverged a lot and decision was to have a separate fix for 4.15 alone and 4.16 will be a separate one. That's what I remember.

leelavg commented:
@rohan47 thinking again, I agree w/ Nitin, we might want to go with checking the package name, able to use existing labels set by olm seems good but it has edge cases.

Btw, pls do update necessary rbac.

rohan47 commented:
Testing the changes

rohan47 commented:
/bugzilla refresh

Madhu-1 commented on controllers/clusterversion_controller.go:
```diff
@@ -104,6 +108,7 @@ func (c *ClusterVersionReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&configv1.ClusterVersion{}, clusterVersionPredicates).
 		Watches(&corev1.ConfigMap{}, enqueueClusterVersionRequest, configMapPredicates).
+		Watches(&opv1a1.Subscription{}, &handler.EnqueueRequestForObject{}).
```
Add a predicate to trigger reconcile for only subscription in operator namespace

Madhu-1 commented on controllers/clusterversion_controller.go:
```diff
@@ -436,3 +447,36 @@ func (c *ClusterVersionReconciler) ensureConsolePlugin() error {
+	)
+
+	for i := 0; i < len(subscriptions.Items); i++ {
+		if subscriptions.Items[i].Spec.Package == csiAddonsSubscriptionPackageName {
+			csiAddonsSubscription = &subscriptions.Items[i]
+		}
+		if subscriptions.Items[i].Spec.Package == ocsClientOperatorSubscriptionPackageName {
+			ocsClientOperatorSubscription = &subscriptions.Items[i]
+		}
```
what if we dont find these subscriptions? can we sync and have common idea for this PR and https://github.com/red-hat-storage/ocs-client-operator/pull/159 as the internal on how we are handling the subscription remains same?

Madhu-1 commented on controllers/clusterversion_controller.go:
```diff
@@ -436,3 +465,40 @@ func (c *ClusterVersionReconciler) ensureConsolePlugin() error {
+	}
+
+	//We want the channel of csiAddons subscription to be stable-4.15 so no need to compare it with ocs-client-operator subscription
+	csiAddonsSubscription.Spec.Channel = "stable-4.15"
```
you need to do update here which is missing

rohan47 replied:
Added

leelavg replied:
Comment should be slightly rephrased, the reasoning we want stable-4.15 is because of a direct merge to release-4.15 channel and this code shouldn't be back ported and from 4.16 anyways the logic is already changed.

rohan47 replied:
Updated the comment to
```
//the channel of csiAddons subscription here is stable-4.15 because of a direct merge to release-4.15
//this code shouldn't be back ported and from release-4.16 the logic is changed
```

leelavg commented:
/bugzilla refresh

leelavg commented on controllers/clusterversion_controller.go:
```diff
@@ -436,3 +465,44 @@ func (c *ClusterVersionReconciler) ensureConsolePlugin() error {
+		return fmt.Errorf("failed to update csi-addons subscription: %v", err)
+	}
+
+	if err != nil {
```
what is this check for? seems redundant to me.

rohan47 replied:
removed

Madhu-1 commented on controllers/clusterversion_controller.go:
```diff
@@ -436,3 +465,41 @@ func (c *ClusterVersionReconciler) ensureConsolePlugin() error {
+
+	}
+
+	//the channel of csiAddons subscription here is stable-4.15 because of a direct merge to release-4.15
+	//this code shouldn't be back ported and from release-4.16 the logic is changed
```
```suggestion
// The channel of csiAddons subscription here is stable-4.15 because of a direct merge to release-4.15. This code shouldn't be backported, and from release-4.16, the logic has changed.
```

add a space between `//` and the word

rohan47 replied:
done

Madhu-1 reviewed (commented):
small nit, LGTM.

rohan47 commented:
Getting this on local

```$ make lint
docker run --rm -v /home/mrudraia/redhat/ocs-client-operator:/app:Z -w /app golangci/golangci-lint:v1.54.2 golangci-lint run ./...
Unable to find image 'golangci/golangci-lint:v1.54.2' locally
v1.54.2: Pulling from golangci/golangci-lint
de4cac68b616: Pull complete 
d31b0195ec5f: Pull complete 
9b1fd34c30b7: Pull complete 
2f744fb50e06: Pull complete 
661ee9827c7b: Pull complete 
fa5c1764d694: Pull complete 
824755c07d59: Pull complete 
1edf93b66cce: Pull complete 
Digest: sha256:2082f5379c48c46e447bc1b890512f3aa9339db5eeed1a483a34aae9476ba6ee
Status: Downloaded newer image for golangci/golangci-lint:v1.54.2
level=warning msg="[runner] The linter 'deadcode' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused."
level=warning msg="[runner] The linter 'structcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused."
level=warning msg="[runner] The linter 'varcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused."```

Madhu-1 commented:
>  level=error msg="[linters_context] deadcode: This linter is fully inactivated: it will not produce any reports."
  level=error msg="[linters_context] structcheck: This linter is fully inactivated: it will not produce any reports."
  level=error msg="[linters_context] varcheck: This linter is fully inactivated: it will not produce any reports."

@rohan47 can you update the ci-lint from v2 to v4 and see if that works?

rohan47 commented:
> > level=error msg="[linters_context] deadcode: This linter is fully inactivated: it will not produce any reports."
> > level=error msg="[linters_context] structcheck: This linter is fully inactivated: it will not produce any reports."
> > level=error msg="[linters_context] varcheck: This linter is fully inactivated: it will not produce any reports."
> 
> @rohan47 can you update the ci-lint from v2 to v4 and see if that works?

Where do I update ci-lint from v2 to v4?
Do I update the golint image tag in `hack/make-project-vars.mk:GO_LINT_IMG_TAG ?= v1.54.2`

Madhu-1 commented:
> > > level=error msg="[linters_context] deadcode: This linter is fully inactivated: it will not produce any reports."
> > > level=error msg="[linters_context] structcheck: This linter is fully inactivated: it will not produce any reports."
> > > level=error msg="[linters_context] varcheck: This linter is fully inactivated: it will not produce any reports."
> > 
> > 
> > @rohan47 can you update the ci-lint from v2 to v4 and see if that works?
> 
> Where do I update ci-lint from v2 to v4? Do I make the v2 to v4 change as per https://github.com/red-hat-storage/ocs-client-operator/blob/main/.github/workflows/golangci-lint.yaml Do I update the golint image tag in `hack/make-project-vars.mk:GO_LINT_IMG_TAG ?= v1.54.2`

add a commit to change v2 to v4 above, i don't think GO_LINT_IMG_TAG is used in CI to install specific version of golangci-lint

leelavg commented:
Hmm, we need to do some more housekeeping then. CI seems to be using latest lint always irrespective of Makefile vars or release branch.

Maybe it's better to use a specific version for branch and latest in main branch 🤔

rohan47 commented:
> > > > level=error msg="[linters_context] deadcode: This linter is fully inactivated: it will not produce any reports."
> > > > level=error msg="[linters_context] structcheck: This linter is fully inactivated: it will not produce any reports."
> > > > level=error msg="[linters_context] varcheck: This linter is fully inactivated: it will not produce any reports."
> > > 
> > > 
> > > @rohan47 can you update the ci-lint from v2 to v4 and see if that works?
> > 
> > 
> > Where do I update ci-lint from v2 to v4? Do I make the v2 to v4 change as per https://github.com/red-hat-storage/ocs-client-operator/blob/main/.github/workflows/golangci-lint.yaml Do I update the golint image tag in `hack/make-project-vars.mk:GO_LINT_IMG_TAG ?= v1.54.2`
> 
> add a commit to change v2 to v4 above, i don't think GO_LINT_IMG_TAG is used in CI to install specific version of golangci-lint

Done, can you please approve the workflow

leelavg commented:
Gave a try locally, first time it failed similar to CI but later it got succeeded. @rohan47 pls add `-v` as well to https://github.com/red-hat-storage/ocs-client-operator/blob/release-4.15/.github/workflows/golangci-lint.yaml#L42 and let's se if CI shows any extra failures.

rohan47 commented:
> Hmm, we need to do some more housekeeping then. CI seems to be using latest lint always irrespective of Makefile vars or release branch.
> 
> Maybe it's better to use a specific version for branch and latest in main branch 🤔

Added `-v` in the args and based on the following 
```Warning: Unexpected input(s) 'skip-go-installation', valid inputs are ['version', 'args', 'working-directory', 'github-token', 'only-new-issues', 'skip-cache', 'skip-pkg-cache', 'skip-build-cache', 'install-mode']```
removed `skip-go-installation`

Madhu-1 commented on .golangci.yaml:
```diff
@@ -7,12 +7,9 @@ run:
 linters:
   disable-all: true
   enable:
-    - deadcode
```
@rohan47 it looks like the changes in this file are enough to pass the CI, can you please revert the changes in the golangcilint.yml file in action?

rohan47 replied:
done

Madhu-1 commented:
> Hmm, we need to do some more housekeeping then. CI seems to be using latest lint always irrespective of Makefile vars or release branch.
> 
> Maybe it's better to use a specific version for branch and latest in main branch 🤔

Yes i think we should stick to versions instead of depending on the latest version which is not meant for release branches

Madhu-1 commented:
@leelavg if everything looks good, please place an `approve` comment from your side.

leelavg commented:
/lgtm
/approve

Finally, I was waiting for CI.

Madhu-1 commented:
/approve
